javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -0,0 +1,209 @@
+        return new BitMappedTrie<>(root, offset, length(), depthShift);
+    }
+
+    BitMappedTrie<T> drop(int n) {
+        if (n <= 0) {
+            return this;
+        } else if (n >= length()) {
+            return empty();
+        } else {
+            final NodeModifier drop = Arrays::copyDrop;
+            final Object[] root = modifyLeaf(array, depthShift, offset + n, drop, drop);
+            return new BitMappedTrie<>(root, offset + n, length() - n, depthShift);
+        }
+    }
+
+    BitMappedTrie<T> take(int n) {
+        if (n >= length()) {
+            return this;
+        } else if (n <= 0) {
+            return empty();
+        } else {
+            final NodeModifier take = Arrays::copyTake;
+            final Object[] root = modifyLeaf(array, depthShift, (offset + n) - 1, take, take);
+            return new BitMappedTrie<>(root, offset, n, depthShift);
+        }
+    }
+
+    /* descend the tree from root to leaf, applying the given modifications along the way, returning the new root */
+    private static Object[] modifyLeaf(Object[] root, int depthShift, int index, NodeModifier node, NodeModifier leaf) {
+        if (depthShift == 0) {
+            return leaf.apply(root, index);
+        } else {
+            int previousIndex = firstDigit(index, depthShift);
+            root = node.apply(root, previousIndex);
+
+            Object[] array = root;
+            for (int shift = depthShift - BRANCHING_BASE; shift >= BRANCHING_BASE; shift -= BRANCHING_BASE) {
+                final int offset = digit(index, shift);
+
+                final Object[] newNode = node.apply(array[previousIndex], offset);
+                array[previousIndex] = newNode;
+
+                previousIndex = offset;
+                array = newNode;
+            }
+
+            array[previousIndex] = leaf.apply(array[previousIndex], lastDigit(index));
+            return root;
+        }
+    }
+
+    T get(int index) {
+        final T[] leaf = getLeaf(index);
+        final int leafIndex = lastDigit(offset() + index);
+        return leaf[leafIndex];
+    }
+
+    /**
+     * fetch the leaf, corresponding to the given index.
+     * Node: the offset and length should be taken into consideration as there may be leading and trailing garbage.
+     */
+    @SuppressWarnings("unchecked")
+    T[] getLeaf(int index) {
+        Object[] leaf = array;
+        for (int shift = depthShift; shift > 0; shift -= BRANCHING_BASE) {
+            leaf = getAt(leaf, digit(offset + index, shift));
+        }
+        return (T[]) leaf;
+    }
+
+    Iterator<T> iterator() {
+        return new Iterator<T>() {
+            int globalIndex = 0;
+            int index = lastDigit(offset());
+            T[] leaf = getLeaf(globalIndex);
+            int length = leaf.length;
+
+            @Override
+            public boolean hasNext() { return globalIndex < BitMappedTrie.this.length(); }
```

Comment:
@danieldietrich, you're checking the bytecode, not the JITted result.
If you're interested whether they're inlined or not, create a JMH test, enable print inlining and run the tests :). But then again, it may be JDK dependent ... maybe even processor dependent.

> Please use globalIndex < length here.

Ok, I will, but I though you asked me to do optimizations in another commit.